bughouse_console/src/async_server_main.rs

```diff
@@ -15,29 +15,47 @@ use rand::RngCore;
 use tide::StatusCode;
 use tungstenite::protocol;
 
-use bughouse_chess::server::*;
 use bughouse_chess::server_hooks::ServerHooks;
+use bughouse_chess::{server::*, BughouseServerEvent};
 use bughouse_console::auth;
 use bughouse_console::database;
 use bughouse_console::http_server_state::*;
-use bughouse_console::session::Session;
+use bughouse_console::session_store::*;
 
 use crate::auth_handlers_tide::*;
 use crate::network::{self, CommunicationError};
 use crate::server_main::{AuthOptions, DatabaseOptions, ServerConfig, SessionOptions};
 use crate::sqlx_server_hooks::*;
 
-async fn handle_connection<S: AsyncRead + AsyncWrite + Unpin + Send + Sync + 'static>(
+async fn handle_connection<DB, S: AsyncRead + AsyncWrite + Unpin + Send + Sync + 'static>(
     peer_addr: String,
     stream: WebSocketStream<S>,
     tx: mpsc::SyncSender<IncomingEvent>,
     clients: Arc<Mutex<Clients>>,
-    session: Option<Session>,
+    session_id: Option<SessionId>,
+    http_server_state: HttpServerState<DB>,
 ) -> tide::Result<()> {
     let (mut stream_tx, mut stream_rx) = stream.split();
-    info!("Client connected: {}, session={:?}", peer_addr, session);
+    info!("Client connected: {}", peer_addr);
 
     let (client_tx, client_rx) = mpsc::channel();
+
+    let mut session_store_subscription_id = None;
+    if let Some(session_id) = &session_id {
+        // Subscribe the client to all updates to the session in session store.
+        let my_client_tx = client_tx.clone();
+        session_store_subscription_id = Some(http_server_state.session_store.lock().unwrap().subscribe(
+            session_id,
+            move |s| {
+                // Send the entire session data to the client.
+                // We can perform some mapping here if we want to hide
+                // some of the state from the client.
+                let _ =
```

Comment:
As far as I understand, unwrapping `send` error is the right thing to do in many cases:
https://doc.rust-lang.org/std/sync/mpsc/#disconnection
For example, we do so 
https://github.com/amatveiakin/bughouse-chess/blob/473a9a2733d0203272a473c45cb6a0f342572ba8/src/server.rs#L149

Do I understand correctly that it's not safe here, because session handling is done asynchronously and we cannot ensure that `client_rx` outlives the subscription?

Reply:
Yes, I could move the cleanup into the thread that owns client_rx but that's a bit subtle.